cli/azd/pkg/project/service_target_aks.go

```diff
@@ -0,0 +1,444 @@
+			environment.AksClusterEnvVarName,
+		)
+	}
+
+	// Login to container registry.
+	loginServer, has := t.env.Values[environment.ContainerRegistryEndpointEnvVarName]
+	if !has {
+		return ServiceDeploymentResult{}, fmt.Errorf(
+			"could not determine container registry endpoint, ensure %s is set as an output of your infrastructure",
+			environment.ContainerRegistryEndpointEnvVarName,
+		)
+	}
+
+	log.Printf("getting AKS credentials %s\n", clusterName)
+	progress <- "Getting AKS credentials"
+	credentials, err := t.containerService.GetAdminCredentials(ctx, t.scope.ResourceGroupName(), clusterName)
+	if err != nil {
+		return ServiceDeploymentResult{}, fmt.Errorf("failed retrieving cluster admin credentials, %w", err)
+	}
+
+	kubeConfigManager, err := kubectl.NewKubeConfigManager(t.kubectl)
+	if err != nil {
+		return ServiceDeploymentResult{}, err
+	}
+
+	kubeConfig, err := kubectl.ParseKubeConfig(ctx, credentials.Kubeconfigs[0].Value)
+	if err != nil {
+		return ServiceDeploymentResult{}, fmt.Errorf("failed parsing kube config: %w", err)
+	}
+
+	if err := kubeConfigManager.SaveKubeConfig(ctx, clusterName, kubeConfig); err != nil {
+		return ServiceDeploymentResult{}, fmt.Errorf("failed saving kube config: %w", err)
+	}
+
+	if err := kubeConfigManager.MergeConfigs(ctx, "config", "config", clusterName); err != nil {
+		return ServiceDeploymentResult{}, fmt.Errorf("failed merging kube configs: %w", err)
+	}
+
+	if _, err := t.kubectl.ConfigUseContext(ctx, clusterName, nil); err != nil {
+		return ServiceDeploymentResult{}, fmt.Errorf("failed using kube context '%s', %w", clusterName, err)
+	}
+
+	kubeFlags := kubectl.KubeCliFlags{
+		Namespace: namespace,
+		DryRun:    "client",
+		Output:    "yaml",
+	}
+
+	progress <- "Creating k8s namespace"
+	namespaceResult, err := t.kubectl.CreateNamespace(
+		ctx,
+		namespace,
+		&kubectl.KubeCliFlags{DryRun: "client", Output: "yaml"},
+	)
+	if err != nil {
+		return ServiceDeploymentResult{}, fmt.Errorf("failed creating kube namespace: %w", err)
+	}
+
+	_, err = t.kubectl.ApplyPipe(ctx, namespaceResult.Stdout, nil)
+	if err != nil {
+		return ServiceDeploymentResult{}, fmt.Errorf("failed applying kube namespace: %w", err)
+	}
+
+	progress <- "Creating k8s secrets"
+	secretResult, err := t.kubectl.CreateSecretGenericFromLiterals(ctx, "azd", t.env.Environ(), &kubeFlags)
+	if err != nil {
+		return ServiceDeploymentResult{}, fmt.Errorf("failed setting kube secrets: %w", err)
+	}
+
+	_, err = t.kubectl.ApplyPipe(ctx, secretResult.Stdout, nil)
+	if err != nil {
+		return ServiceDeploymentResult{}, fmt.Errorf("failed applying kube secrets: %w", err)
+	}
+
+	log.Printf("logging into registry %s\n", loginServer)
+
+	progress <- "Logging into container registry"
+	if err := t.az.LoginAcr(ctx, t.docker, t.env.GetSubscriptionId(), loginServer); err != nil {
+		return ServiceDeploymentResult{}, fmt.Errorf("failed logging into registry '%s': %w", loginServer, err)
+	}
```

Comment:
nit. Move the `login to Arc` and any other operation which might fail at the middle of the flow to be an initial validation for fast-fail.
It will be frustrating to fail after some operations just because there's a problem with the login